plenum/common/channel.py

```diff
@@ -90,20 +120,68 @@ def put_nowait(self, msg: Any) -> bool:
     def __init__(self):
         self._queue = Queue()
         self._inbox = self.Channel(self._queue)
-        self._router = _RouterRxChannel()
+        self._observable = _RxChannel()
         self._is_running = False
 
     def inbox(self) -> TxChannel:
         return self._inbox
 
     def router(self) -> RxChannel:
-        return self._router
+        return self._observable
 
     async def run(self):
         self._is_running = True
         while self._is_running:
             msg = await self._queue.get()
-            await self._router.handle(msg)
+            await self._observable.notify(msg)
 
     def stop(self):
         self._is_running = False
+
+
+class RouterBase:
+    def __init__(self, strict: bool = True):
+        self._strict = strict
+        self._routes = {}  # type: Dict[Type, Callable]
+
+    def add(self, msg_type: Type, handler: Callable):
+        self._routes[msg_type] = handler
+
+    def _process_sync(self, msg: Any):
+        # This is done so that messages can include additional metadata
+        # isinstance is not used here because it returns true for NamedTuple
+        # as well.
+        if type(msg) != tuple:
+            msg = (msg,)
+        handler = self._find_handler(msg[0])
+        if handler is None:
+            if self._strict:
+                raise RuntimeError("unhandled msg: {}".format(msg))
+            return
+        return handler(*msg)
+
+    def _find_handler(self, msg: Any) -> Callable:
+        for cls, handler in self._routes.items():
+            if isinstance(msg, cls):
+                return handler
+
+
+class Router(RouterBase):
+    def __init__(self, input: RxChannel, strict: bool = True):
+        RouterBase.__init__(self, strict)
+        input.subscribe(self._process_sync)
+
+    def add(self, msg_type: Type, handler: Callable):
+        if iscoroutinefunction(handler):
+            raise ValueError('Router works only with synchronous handlers')
+        RouterBase.add(self, msg_type, handler)
+
+
+class AsyncRouter(RouterBase):
```

Comment:
It would be great to have tests for all the classes from this module

Reply:
The tests will make sure that everything works as intended. Also they can be used as examples.